S3 Java SDK copies bytebuffers to the heap - can cause OOM

Seems to be an issue with the chunkEncoding implementation of the AWS request signer.

Solution could be to disable chunkEncoding.

```
2026-02-14T07:00:02.570980433Z xtdb.indexer.LogProcessor ERROR	
Ingestion stopped for 'xtdb' database: error processing log record at id 562949994086779 (epoch: 2, logOffset: 40665467)
java.util.concurrent.ExecutionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Encountered fatal error in publisher (SDK Attempt Count: 1)
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at xtdb.storage.RemoteBufferPool.putObject(RemoteBufferPool.kt:196)
	at xtdb.table_catalog.TableCatalog.finish_block_BANG_(table_catalog.clj:160)
	at xtdb.indexer.live_index.LiveIndex.finishBlock(live_index.clj:158)
	at xtdb.indexer.LogProcessor.finishBlock(LogProcessor.kt:307)
	at xtdb.indexer.LogProcessor$processRecords$1.invokeSuspend(LogProcessor.kt:270)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:263)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:94)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:70)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:48)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at xtdb.indexer.LogProcessor.processRecords(LogProcessor.kt:154)
	at xtdb.api.log.KafkaCluster$KafkaLog$tailAll$job$1.invokeSuspend$lambda$4$lambda$3(KafkaCluster.kt:281)
	at kotlinx.coroutines.InterruptibleKt.runInterruptibleInExpectedContext(Interruptible.kt:48)
	at kotlinx.coroutines.InterruptibleKt.access$runInterruptibleInExpectedContext(Interruptible.kt:1)
	at kotlinx.coroutines.InterruptibleKt$runInterruptible$2.invokeSuspend(Interruptible.kt:40)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:124)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:820)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Encountered fatal error in publisher (SDK Attempt Count: 1)
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:130)
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:95)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.retryPolicyDisallowedRetryException(RetryableStageHelper.java:168)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeAttemptExecute(AsyncRetryableStage.java:139)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:157)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:117)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:78)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$execute$0(MakeAsyncHttpRequestStage.java:108)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.completeResponseFuture(MakeAsyncHttpRequestStage.java:255)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$3(MakeAsyncHttpRequestStage.java:167)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: Encountered fatal error in publisher
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$ReplayableByteBufferSubscription.processRequest(ByteBuffersAsyncRequestBody.java:239)
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$ReplayableByteBufferSubscription.request(ByteBuffersAsyncRequestBody.java:226)
	at software.amazon.awssdk.utils.async.AddingTrailingDataSubscriber$1.request(AddingTrailingDataSubscriber.java:94)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.ensureUpstreamDemandExists(BaseSubscriberAdapter.java:301)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.handleUpstreamDemandState(BaseSubscriberAdapter.java:230)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.handleStateUpdate(BaseSubscriberAdapter.java:180)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter$1.request(BaseSubscriberAdapter.java:109)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeRequestMore(HandlerSubscriber.java:303)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeStart(HandlerSubscriber.java:238)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.provideSubscription(HandlerSubscriber.java:225)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.access$000(HandlerSubscriber.java:41)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber$1.run(HandlerSubscriber.java:213)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:507)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1073)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	... 1 common frames omitted
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.base/java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:71)
	at java.base/java.nio.ByteBuffer.allocate(ByteBuffer.java:391)
	at software.amazon.awssdk.http.auth.aws.internal.signer.chunkedencoding.ChunkedEncodedPublisher.encodeChunk(ChunkedEncodedPublisher.java:186)
	at software.amazon.awssdk.http.auth.aws.internal.signer.chunkedencoding.ChunkedEncodedPublisher.access$800(ChunkedEncodedPublisher.java:63)
	at software.amazon.awssdk.http.auth.aws.internal.signer.chunkedencoding.ChunkedEncodedPublisher$ChunkingSubscriber.onNext(ChunkedEncodedPublisher.java:318)
	at software.amazon.awssdk.http.auth.aws.internal.signer.chunkedencoding.ChunkedEncodedPublisher$ChunkingSubscriber.onNext(ChunkedEncodedPublisher.java:272)
	at software.amazon.awssdk.utils.async.ContentLengthAwareSubscriber.onNext(ContentLengthAwareSubscriber.java:55)
	at software.amazon.awssdk.utils.async.ContentLengthAwareSubscriber.onNext(ContentLengthAwareSubscriber.java:27)
	at software.amazon.awssdk.http.auth.aws.internal.signer.io.UnbufferedChecksumSubscriber.onNext(UnbufferedChecksumSubscriber.java:54)
	at software.amazon.awssdk.http.auth.aws.internal.signer.io.UnbufferedChecksumSubscriber.onNext(UnbufferedChecksumSubscriber.java:33)
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$ReplayableByteBufferSubscription.doProcessRequest(ByteBuffersAsyncRequestBody.java:271)
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$ReplayableByteBufferSubscription.processRequest(ByteBuffersAsyncRequestBody.java:237)
	at software.amazon.awssdk.core.internal.async.ByteBuffersAsyncRequestBody$ReplayableByteBufferSubscription.request(ByteBuffersAsyncRequestBody.java:226)
	at software.amazon.awssdk.utils.async.AddingTrailingDataSubscriber$1.request(AddingTrailingDataSubscriber.java:94)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.ensureUpstreamDemandExists(BaseSubscriberAdapter.java:301)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.handleUpstreamDemandState(BaseSubscriberAdapter.java:230)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter.handleStateUpdate(BaseSubscriberAdapter.java:180)
	at software.amazon.awssdk.utils.async.BaseSubscriberAdapter$1.request(BaseSubscriberAdapter.java:109)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeRequestMore(HandlerSubscriber.java:303)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.maybeStart(HandlerSubscriber.java:238)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.provideSubscription(HandlerSubscriber.java:225)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber.access$000(HandlerSubscriber.java:41)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerSubscriber$1.run(HandlerSubscriber.java:213)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:507)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1073)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.runWith(Thread.java:1596)
	... 1 common frames omitted
```

--- a/modules/aws/src/main/kotlin/xtdb/aws/S3.kt
+++ b/modules/aws/src/main/kotlin/xtdb/aws/S3.kt
@@ -78,6 +78,14 @@ class S3(
         client.close()
     }
 
+    private fun efficientAsyncRequestBody(buf: ByteBuffer): AsyncRequestBody =
+        buf
+            // prevents copying the buffer into a new heap ByteBuffer, as `fromByteBuffer` would do
+            .let(AsyncRequestBody::fromByteBufferUnsafe)
+            // makes the buffer non-replayable, again preventing heap copies for payload-signing or checksum calculations
+            // (couldn't achive this effect by S3 client configuration)
+            .let(AsyncRequestBody::fromPublisher)
+
     override fun startMultipart(k: Path): CompletableFuture<IMultipartUpload<CompletedPart>> = scope.future {
         val s3Key = prefix.resolve(k).toString()
         val initResp = client.createMultipartUpload {
@@ -100,7 +108,7 @@ class S3(
                     val contentLength = buf.remaining().toLong()
                     val partNum = idx + 1
 
-                    val partResp = client.uploadPart(AsyncRequestBody.fromByteBufferUnsafe(buf)) {
+                    val partResp = client.uploadPart(efficientAsyncRequestBody(buf)) {
                         it.bucket(bucket)
                         it.key(s3Key)
                         it.uploadId(uploadId)
@@ -179,7 +187,7 @@ class S3(
 
             val contentLength = buf.remaining().toLong()
 
-            client.putObject(AsyncRequestBody.fromByteBufferUnsafe(buf)) {
+            client.putObject(efficientAsyncRequestBody(buf)) {
                 it.bucket(bucket)
                 it.key(s3Key)
                 it.contentLength(contentLength)

